core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java

```diff
@@ -638,12 +641,120 @@ public static Integer jsonStorageSize(String input) {
     }
   }
 
+  private static String jsonPairsModify(Object jsonDoc, String type, Object... kvs) {
+    assert kvs.length % 2 == 0;
+    String result = "";
+    Map<String, Object> map = new HashMap<>();
+    DocumentContext ctx = JsonPath.parse(jsonDoc.toString(),
+            Configuration
+                    .builder()
+                    .options(Option.SUPPRESS_EXCEPTIONS)
+                    .jsonProvider(JSON_PATH_JSON_PROVIDER)
+                    .mappingProvider(JSON_PATH_MAPPING_PROVIDER)
+                    .build());
+
+    for (int i = 0; i < kvs.length; i += 2) {
+      String k = (String) kvs[i];
+      Object v = kvs[i + 1];
+      if (JsonPath.isPathDefinite(k) && k.contains(JSON_ROOT_PATH)) {
+        map.put(k, v);
+      } else {
+        throw RESOURCE.validationError(k).ex();
+      }
+    }
+
+    for (Map.Entry<String, Object> entry : map.entrySet()) {
+      String k = entry.getKey();
+      Object v = entry.getValue();
+      switch (type) {
+      case "replace":
+        if (k.equals(JSON_ROOT_PATH)) {
+          result = jsonize(v);
+        } else {
+          if (ctx.read(k) != null) {
+            ctx.set(k, v);
+          }
+        }
+        break;
+      case "insert":
+        if ((!k.equals(JSON_ROOT_PATH)) && (ctx.read(k) == null)) {
+          insertToJson(ctx, k, v);
+        }
+        break;
+      case "set":
+        if (k.equals(JSON_ROOT_PATH)) {
+          result = jsonize(v);
+        } else {
+          if (ctx.read(k) != null) {
+            ctx.set(k, v);
+          } else {
+            insertToJson(ctx, k, v);
+          }
+        }
+        break;
+      }
+    }
+    result = Strings.isNullOrEmpty(result) ?  ctx.jsonString() : result;
+    return result;
+  }
+
+  private static void insertToJson(DocumentContext ctx, String path, Object value) {
+    final String preantPath;
+    final String key;
+    Integer dotIndex = path.lastIndexOf(".");
+    Integer leftBracketIndex = path.lastIndexOf("[");
+    if (dotIndex.equals(-1) && leftBracketIndex.equals(-1)) {
+      preantPath = path;
+      key = path;
+    } else if (!dotIndex.equals(-1) && leftBracketIndex.equals(-1)) {
+      preantPath = path.substring(0, dotIndex);
+      key = path.substring(dotIndex + 1);
+    } else if (dotIndex.equals(-1) && !leftBracketIndex.equals(-1)) {
+      preantPath = path.substring(0, leftBracketIndex);
+      key = path.substring(leftBracketIndex + 1);
+    } else {
+      Integer position = dotIndex > leftBracketIndex ? dotIndex : leftBracketIndex;
+      preantPath = path.substring(0, position);
+      key = path.substring(position);
+    }
+    Object obj = ctx.read(preantPath);
+    if (obj instanceof Map) {
+      ctx.put(preantPath, key, value.toString());
+    } else if (obj instanceof Collection) {
+      ctx.add(preantPath, value);
+    }
+  }
+
+  public static String jsonInsert(Object jsonDoc, Object... kvs) {
```

Comment:
Please add support for string input. We should support both String and JsonValueContext as the first input argument for such methods. Don't use Object. You can take JSON_REMOVE as an example.